### ISSUE TITLE
Parcel memory leak crash after a few HMRs

### ISSUE THREAD

**Issue by natew** [parcel-bundler/parcel#1185]
bug reprot

Basically, parcel 1.6 crashes pretty consistently for me when saving files in a medium-large app that uses HMR.

The stack is as follows:

```
<--- Last few GCs --->

[1273:0x103800000]  3702603 ms: Mark-sweep 1369.2 (1452.7) -> 1368.5 (1460.7) MB, 1416.1 / 0.0 ms  allocation failure GC in old space requested
[1273:0x103800000]  3703648 ms: Mark-sweep 1368.5 (1460.7) -> 1368.4 (1429.7) MB, 1044.4 / 0.0 ms  last resort GC in old space requested
[1273:0x103800000]  3704689 ms: Mark-sweep 1368.4 (1429.7) -> 1368.4 (1429.7) MB, 1040.5 / 0.0 ms  last resort GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x224ac1d25529 <JSObject>
    0: builtin exit frame: stringify(this=0x224ac1d08d59 <Object map = 0x224a7a102ba1>,0x224ab5b022d1 <undefined>,0x224ab5b022d1 <undefined>,0x224a8a01f831 <Object map = 0x224a3f4a9d41>)

    1: arguments adaptor frame: 1->3
    2: toString(aka SourceMapGenerator_toString) [/Users/nw/projects/motion/orbit/node_modules/parcel-bundler/node_modules/source-map/lib/source-map-generator.js:422] [bytec...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/Users/nw/n/bin/node]
 2: node::FatalTryCatch::~FatalTryCatch() [/Users/nw/n/bin/node]
 3: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/Users/nw/n/bin/node]
 4: v8::internal::Factory::NewRawOneByteString(int, v8::internal::PretenureFlag) [/Users/nw/n/bin/node]
 5: v8::internal::String::SlowFlatten(v8::internal::Handle<v8::internal::ConsString>, v8::internal::PretenureFlag) [/Users/nw/n/bin/node]
 6: v8::internal::JsonStringifier::SerializeString(v8::internal::Handle<v8::internal::String>) [/Users/nw/n/bin/node]
 7: v8::internal::JsonStringifier::Result v8::internal::JsonStringifier::Serialize_<true>(v8::internal::Handle<v8::internal::Object>, bool, v8::internal::Handle<v8::internal::Object>) [/Users/nw/n/bin/node]
 8: v8::internal::JsonStringifier::Result v8::internal::JsonStringifier::Serialize_<false>(v8::internal::Handle<v8::internal::Object>, bool, v8::internal::Handle<v8::internal::Object>) [/Users/nw/n/bin/node]
 9: v8::internal::JsonStringifier::Stringify(v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>) [/Users/nw/n/bin/node]
10: v8::internal::Builtin_Impl_JsonStringify(v8::internal::BuiltinArguments, v8::internal::Isolate*) [/Users/nw/n/bin/node]
11: 0x367c4b50697d
12: 0x367c4b50535f
13: 0x367c4b5bd196
14: 0x367c4b5bd196
15: 0x367c4b5bd196
```

### 🎛 Configuration (.babelrc, package.json, cli command)

```js
{
  "your": { "config": "here" }
}
```

### 🤔 Expected Behavior

Not crashing

Unfortunately, this one is hard for me to help you guys with without doing a more intense profile on the running process. The app is complex, and it only happens usually after a few HMRs.

Separately, tried updating to 1.7 but it won't run our stack anymore (forget the error).

### 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 1.6.1
| Node             | 9.8
| npm/Yarn         | yarn 1.5.1
| Operating System | high sierra

**Comment by devongovett** [comment]
How big is your app, e.g. how many files + node_modules?

**Comment by natew** [comment]
Good question, any way to get parcel to spit it out?

**Comment by DeMoorJasper** [comment]
@natew You could use this option https://parceljs.org/cli.html#print-a-detailed-report using a build command to get all the requires and sizes parcel processes.
Although this will probably not be possible as in your case parcel crashes before even reaching that point?

**Comment by natew** [comment]
Ah cool will run that soon. It only crashes after using it for a few HMRs
so should be fine to print.

**Comment by samsartor** [comment]
I also have a failing [travis build](https://travis-ci.org/samsartor/regess/builds/371843432).

Note that running `yarn install` before `parcel build` does not help.

**Comment by rafaelrinaldi** [comment]
This seem to be related to #1513

**Comment by gimenete** [comment]
I've had the same problem and my workaround has been to use `--no-cache`. It seems to work now.

**Comment by alidcastano** [comment]
I have a similar problem. Along with random `Object prototype may only be an Object or null: undefined` errors (#1181), I'm getting warnings about memory leaks, e.g. `MaxListenersExceededWarning: Possible EventEmitter memory leak detected.` So sometimes my app starts, sometimes it doesn't.

**Comment by mzabsky** [comment]
Confirming this issue. Just got:

```
<--- Last few GCs --->[24344:0000028F681B9D00] 93613091 ms: Mark-sweep 1369.8 (1432.4) -> 1369.8 (1435.4) MB, 901.3 / 0.0 ms  allocation failure GC in old space requested
[24344:0000028F681B9D00] 93613962 ms: Mark-sweep 1369.8 (1435.4) -> 1369.7 (1403.4) MB, 870.8 / 0.0 ms  last resort GC in old space requested
[24344:0000028F681B9D00] 93614848 ms: Mark-sweep 1369.7 (1403.4) -> 1369.7 (1403.4) MB, 885.0 / 0.0 ms  last resort GC in old space requested

<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 000000346C325879 <JSObject>
    0: builtin exit frame: stringify(this=000000346C3090A9 <Object map = 0000004EF4082BA1>,000000449D9022D1 <undefined>,000000449D9022D1 <undefined>,000002A827B95B21 <Object map = 000002F0E80A6A61>)

    1: arguments adaptor frame: 1->3
    2: toString(aka SourceMapGenerator_toString) [C:\Users\matej.zabsky\Desktop\lung\node_modules\parcel-bundler\node_modules\source-map\lib\source-map-generat...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node_module_register
 2: v8::internal::FatalProcessOutOfMemory
 3: v8::internal::FatalProcessOutOfMemory
 4: v8::internal::Factory::NewRawOneByteString
 5: v8::internal::Smi::SmiPrint
 6: v8::internal::StackGuard::HandleInterrupts
 7: v8::internal::wasm::LocalDeclEncoder::Size
 8: v8::internal::wasm::LocalDeclEncoder::Size
 9: v8::internal::wasm::LocalDeclEncoder::Size
10: v8_inspector::protocol::Debugger::API::SearchMatch::fromJSONString
11: v8_inspector::protocol::Debugger::API::SearchMatch::fromJSONString
12: 00000390E0686B21
error Command failed with exit code 3.
```

While I am not getting this particular message too consistently, parcel builds consistenly slow down over time when using HMR.  This becomes noticeable as soon as after 15 min of normal development. It is not a big issue though, restarting parcel takes 10 seconds.

My project is a React+Redux app, it has 773 folders in node_modules totaling about 100 MB. I can provide more info, if needed, but I cannot share the complete source code.

`--no-cache` helps, but slows down development much more than just restarting parcel every now and then.

**Comment by DeMoorJasper** [comment]
@mzabsky can you try out the latest master branch using git. It might have been related to this: #1755 as the stacktrace mentions sourcemaps

**Comment by kennetpostigo** [comment]
I reproduce this on master as well with a really long json file.

**Comment by DeMoorJasper** [comment]
It might be because we inline all the sources inside the sourcemap. Can you check using `--no-source-maps`. I don't think this is really a memory leak. It might be just how sourcemaps works (the library) and as the JS version has been depricated, we might have to branch of their project to create a streaming stringifier or something idk.

**Comment by kennetpostigo** [comment]
@DeMoorJasper disabled source maps and it still seems to happen. Is there anyway to tell parcel "don't bundle this, I know what I'm doing"?

**Comment by DeMoorJasper** [comment]
@kennetpostigo not sure, in most stacktraces I saw in this issue it occurs in sourcemap-generator, so it is probably related to sourcemaps for those issues. If you could post your stacktrace it would help a lot (the one without sourcemaps)

**Comment by vikingmute** [comment]
Hey guys, I ran into the same issue, my files are super simple, it's a regular javascript file and html file, and I just wanna set up a development server with HMR.

I tried 
```bash
parcel index.html --no-cache
```
not working, the error came up after several minutes.

```bash
parcel index.html --no-source-maps
```
not working, same issue.

Parcel: 1.10.3
Node: 8.11.3
MacOS Mojave 10.14

**Comment by andrewjrhill** [comment]
Experiencing the same issue. Codebase has a large number of dependencies which is what I imagine is causing the crash.

**Comment by abdul737** [comment]
> Hey guys, I ran into the same issue, my files are super simple, it's a regular javascript file and html file, and I just wanna set up a development server with HMR.
> 
> I tried
> 
> ```shell
> parcel index.html --no-cache
> ```
> 
> not working, the error came up after several minutes.
> 
> ```shell
> parcel index.html --no-source-maps
> ```
> 
> not working, same issue.
> 
> Parcel: 1.10.3
> Node: 8.11.3
> MacOS Mojave 10.14

This worked for me:
> ```shell
> parcel index.html --no-source-maps
> ```

I was including my own custom npm lib which was the size of more than 4MB build, and in both projects I used parcel. Now I have no error. Thanks

**Comment by embeddedt** [comment]
I was running into a similar issue with a moderately large Parcel project. Parcel would consistently eat all of my RAM (6GB) whenever I started it.

After reading through many issues I found this one. Using @DeMoorJasper's suggestion (`--no-source-maps`) seems to have entirely fixed the problem, though at the cost of easily viewing stack traces.

Is there a way to enable source maps just for my code and not for any code inside `node_modules`? That way the final source map would be much smaller.

**Comment by DeMoorJasper** [comment]
@embeddedt unfortunately not.

Parcel 2 has a lot of improvements when it comes to ram usage, including only reading sourcefiles when needed for sourcemaps, which is when the sourcemaps get written if inline-sources is enabled which is only the default in production as we now mount the sources in the dev server. So it should be fixed for Parcel 2. Unfortunately there's no workaround for Parcel 1 besides turning of sourcemaps

**Comment by embeddedt** [comment]
I see. That's too bad.

Is Parcel 2 stable enough yet for production? I noticed that public testing opened up a few days ago.

**Comment by DeMoorJasper** [comment]
@embeddedt it's not stable enough to use in production. But any feedback and bugs discovered on larger codebases are very welcome

**Comment by embeddedt** [comment]
It turns out that VSCode's function name autocompletion had inadvertently added an `import` statement for a very large module which I never used anywhere in my project. Once I removed that module, I was able to start using source maps again with little to no RAM impact.

So my suggestion is: for those who are still using Parcel 1, check your bundles using `source-map-analyzer` or `parcel build --detailed-report` to see if you are importing some large module which you don't need. If you are, removing it will probably fix the issue.
